pkix/cert_auth.go

```diff
@@ -117,12 +164,81 @@ func CreateIntermediateCertificateAuthority(crtAuth *Certificate, keyAuth *Key,
 	return NewCertificateFromDER(crtOutBytes), nil
 }
 
+// CreateIntermediateCertificateAuthorityWithOptions creates an intermediate with options.
+// CA certificate signed by the given authority.
+func CreateIntermediateCertificateAuthorityWithOptions(crtAuth *Certificate, keyAuth *Key, csr *CertificateSigningRequest, proposedExpiry time.Time, opts ...Option) (*Certificate, error) {
+	authTemplate := newAuthTemplate()
+
+	serialNumberLimit := new(big.Int).Lsh(big.NewInt(1), 128)
+	serialNumber, err := rand.Int(rand.Reader, serialNumberLimit)
+	if err != nil {
+		return nil, err
+	}
+	authTemplate.SerialNumber.Set(serialNumber)
+
+	rawCsr, err := csr.GetRawCertificateSigningRequest()
+	if err != nil {
+		return nil, err
+	}
+
+	authTemplate.RawSubject = rawCsr.RawSubject
+
+	caExpiry := time.Now().Add(crtAuth.GetExpirationDuration())
+	// ensure cert doesn't expire after issuer
+	if caExpiry.Before(proposedExpiry) {
+		authTemplate.NotAfter = caExpiry
+	} else {
+		authTemplate.NotAfter = proposedExpiry
+	}
+
+	authTemplate.SubjectKeyId, err = GenerateSubjectKeyID(rawCsr.PublicKey)
+	if err != nil {
+		return nil, err
+	}
+
+	authTemplate.IPAddresses = rawCsr.IPAddresses
+	authTemplate.DNSNames = rawCsr.DNSNames
+	authTemplate.URIs = rawCsr.URIs
+
+	rawCrtAuth, err := crtAuth.GetRawCertificate()
+	if err != nil {
+		return nil, err
+	}
+
+	applyOptions(&authTemplate, opts)
+
+	crtOutBytes, err := x509.CreateCertificate(rand.Reader, &authTemplate, rawCrtAuth, rawCsr.PublicKey, keyAuth.Private)
+	if err != nil {
+		return nil, err
+	}
+
+	return NewCertificateFromDER(crtOutBytes), nil
+}
+
+// WithPathlenOption will check if the certificate should have `pathlen` or not.
+func WithPathlenOption(pathlen int, excludePathlen bool) func(template *x509.Certificate) {
+	return func(template *x509.Certificate) {
+		template.MaxPathLen = pathlen
+		template.MaxPathLenZero = !excludePathlen
```

Comment:
We can unconditionally set `MaxPathLenZero = true` (which is probably better than relying on the default behavior of `newAuthTemplate`. We never need to set it to false if we use `-1` as the "unset" value like you're doing below. To be more explicit:

In the explicit zero case, we'll have:

```
MaxPathLen = 0
MaxPathLenZero = true
```

In the unset case:

```
MaxPathLen = -1
MaxPathLenZero = true
```

And in the nonzero case:

```
MaxPathLen = N
MaxPathLenZero = true
```

This is safe to do because `MaxPathLenZero` is ignored unless `MaxPathLen` is zero: https://cs.opensource.google/go/go/+/refs/tags/go1.18.3:src/crypto/x509/x509.go;l=1305

Reply:
Got it, fixed!